src/Common/src/System/Security/Cryptography/DerEncoder.cs

```diff
@@ -149,6 +149,123 @@ internal static byte[][] SegmentedEncodeUnsignedInteger(byte[] bigEndianBytes, i
+            // should be { 00 01 }
+            //
+            // And, if 8 bits are defined, and only bit 6 is set, instead of { 00 02 } it should be { 01 02 },
+            // signifiying that the last bit was omitted.
+
+            int lastSetBit = -1;
+
+            int lastBitProvided = (bigEndianBytes.Length * 8) - 1;
+            int lastPossibleBit = Math.Min(lastBitProvided, namedBitsCount - 1);
+
+            for (int currentBit = lastPossibleBit; currentBit >= 0; currentBit--)
+            {
+                int currentByte = currentBit / 8;
+
+                // currentBit 0 => currentByte 0, shiftIndex 7 (1 << 7)
+                // currentBit 1 => currentByte 0, shiftIndex 6 (1 << 6)
+                // currentBit 7 => currentByte 0, shiftIndex 0 (1 << 0)
+                // currentBit 8 => currentByte 1, shiftIndex 7 (1 << 7)
+                // etc
+                int shiftIndex = 7 - (currentBit % 8);
+                int testValue = 1 << shiftIndex;
+                byte dataByte = bigEndianBytes[currentByte];
+
+                if ((dataByte & testValue) == testValue)
+                {
+                    lastSetBit = currentBit;
+                    break;
+                }
+            }
+
+            byte[] dataSegment;
+
+            if (lastSetBit >= 0)
+            {
+                // Bits are zero-indexed, so lastSetBit=0 means "1 semantic bit", and
+                // "1 semantic bit" requires a byte to write it down.
+                int semanticBits = lastSetBit + 1;
+                int semanticBytes = (7 + semanticBits) / 8;
+
+                // lastSetBit 0 => mask 0b10000000
+                // lastSetBit 1 => mask 0b11000000
+                // lastSetBit 7 => mask 0b11111111
+                // lastSetBit 8 => mask 0b10000000
+                byte lastByteSemanticMask = unchecked((byte)(-1 << (7 - (lastSetBit % 8))));
+
+                // For a lastSetBit of  : 0 1 2 3 4 5 6 7 8 9 A B C D E F
+                // unused bits should be: 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+                int unusedBits = 7 - (lastSetBit % 8);
+
+                // Semantic bytes plus the "how many unused bits" byte.
+                dataSegment = new byte[semanticBytes + 1];
+                dataSegment[0] = (byte)unusedBits;
+                
+                Debug.Assert(semanticBytes <= bigEndianBytes.Length);
+
+                Buffer.BlockCopy(bigEndianBytes, 0, dataSegment, 1, semanticBytes);
+
+                // But the last byte might have too many bits set, trim it down
+                // to only the ones we knew about (all "don't care" values must be 0)
+                dataSegment[semanticBytes] &= lastByteSemanticMask;
+            }
+            else
+            {
+                // No bits being set is encoded as just "no unused bits",
+                // with no semantic payload.
+                dataSegment = new byte[] { 0x00 };
+            }
+
+            return new[]
```

Comment:
Fair enough, explicitly specified `new byte[][]` here, and `new byte[]` in the first entry.